src/components/PopupWithForm.js

```diff
@@ -22,28 +24,34 @@ export default class PopupWithForm extends Popup {
   }
 
   open() {
-    this.handleLoading(false);
+    this.renderLoading(false);
     super.open();
   }
 
   _getInputValues() {
-    const inputList = [...this._popupForm.querySelectorAll(".popup__input")];
     const inputValues = {};
-    for (const input of inputList) {
+    for (const input of this._inputList) {
       inputValues[input.name] = input.value;
     }
     return inputValues;
   }
 
+  setInputValues(data) {
+    this._inputList.forEach((input) => {
+      input.value = data[input.name];
+    })
+  }
+
 
   setEventListeners() {
     super.setEventListeners();
     this._popupForm.addEventListener("submit", (evt) => {
       evt.preventDefault();
-      this.handleLoading(true);
+      this.renderLoading(true);
       this._handleFormSubmit(this._getInputValues())
-        .then(() => {
+        .finally(() => {
           this.close();
```

Comment:
`this.close();` should be in `then`


```js
      this.renderLoading(true);
      this._handleFormSubmit(this._getInputValues())
        .then(() => {
          this.close();  // close in then
        })
        .finally(() => {
          this.renderLoading(false);
        });

```

You need to close popups only in blocks `then` after a **successful response** from the server so that when there is an error it would be opened (the inputs will not be cleared and the user could send the data again to the server) and if you change the button text you could see the change as well